sql-plugin/src/main/scala/com/nvidia/spark/rapids/SplitColumnarBatchProcessor.scala

```diff
@@ -0,0 +1,333 @@
+   *
+   * Algorithm:
+   * - Initialize `sizeOfBatch` and `resultBatches`.
+   * - Iterate through `partRows`:
+   *   - Add (rows, [values]) to the current batch.
+   *   - For each column:
+   *     - Calculate the size of the current partition for the column.
+   *     - Check if adding this partition to the current batch would exceed cuDF limit.
+   *       - If it does, split the partition:
+   *         - Update the current batch's last value = (new rows, [values]).
+   *         - Add it to the result batches.
+   *         - Process the remaining rows again.
+   *       - Otherwise, increase the `sizeOfBatch` for the column.
+   *
+   * Example:
+   * Input:
+   *  partition rows:   [10, 40, 70, 10, 11]
+   *  partition values: [ [abc, ab], [bc, ab], [abc, bc], [aa, cc], [ade, fd] ]
+   *  cuDF limit:  300 bytes
+   *
+   * Result:
+   * [
+   *  [ (10, [abc, ab]), (40, [bc, ab]), (63, [abc, bc]) ],
+   *  [ (7, [abc, bc]), (10, [aa, cc]), (11, [ade, fd]) ]
+   * ]
+   *
+   * @return An array of batches, containing (row counts, partition values) pairs,
+   *         such that each batch's size is less than cudfLimit.
+   */
+  private def splitPartitionDataIntoGroups(partRows: Array[Long], partValues: Array[InternalRow],
+      partSchema: StructType): Array[Array[(Long, InternalRow)]] = {
+    val resultBatches = ArrayBuffer[Array[(Long, InternalRow)]]()
+    val currentBatch = ArrayBuffer[(Long, InternalRow)]()
+    // Initialize a HashMap to store the size of the current batch for each column
+    val sizeOfBatch = mutable.HashMap.empty[Int, Long]
+    var partIndex = 0
+    while (partIndex < partRows.length) {
+      val rowsInPartition = partRows(partIndex)
+      val valuesInPartition = partValues(partIndex)
+      // Add the partition's data to the current batch
+      currentBatch.append((rowsInPartition, valuesInPartition))
+      partSchema.zipWithIndex.foreach {
+        case (field, colIndex) if field.dataType == StringType =>
+          val singleValue = valuesInPartition.getString(colIndex)
+          val sizeOfSingleValue = singleValue.getBytes("UTF-8").length.toLong
+          val sizeOfPartition = rowsInPartition * sizeOfSingleValue
+          // Check if adding the partition to the current batch exceeds cuDFLimit
+          if (sizeOfBatch.getOrElseUpdate(colIndex, 0L) + sizeOfPartition <= cuDF_LIMIT) {
+            sizeOfBatch(colIndex) += sizeOfPartition
+          } else {
+            // Need to split the current partition
+            val newRows = (cuDF_LIMIT - sizeOfBatch(colIndex)) / sizeOfSingleValue
+            // Update batches and clear states
+            currentBatch.update(currentBatch.length - 1, (newRows, valuesInPartition))
+            resultBatches.append(currentBatch.toArray)
+            currentBatch.clear()
+            sizeOfBatch.clear()
+            // Reprocess the current partition with remaining rows
+            partRows(partIndex) -= newRows
+            partIndex -= 1
```

Comment:
We're in an inner loop that is not incrementing `partIndex`, so won't this potentially cause the next iteration of the `partSchema.zipWithIndex.foreach` to look at the _previous_ partition values rather than the current?